Update draft and leaf details (#4)
ietf-scitt/draft-birkholz-cose-cometre-ccf-profile: draft-birkholz-cose-cometre-ccf-profile.md

---

> title: A Transaction Ledger Verifiable Structure for COSE Merkle Tree Proofs

Reviewer: CCF COSE Receipts?

> title: A Transaction Ledger Verifiable Structure for COSE Merkle Tree Proofs

Author: ```suggestion
title: A CoMETRE Profile for CCF COSE Receipts
```

This title was taken from #3, where it had received positive feedback. What I am suggesting in this comment is what feels most accurate to me, what do you think?

Reviewer: Im not attached to any specific title, but I do prefer shorter titles, and I would avoid adding "CoMETRE" or "Merkle"... "CCF based COSE Receipts" or "COSE Receipts with CCF" seems sufficient.

Author: Updated :) I was lulled into a longer title by the presence of the abbrev field, but I do agree that shorter is better here.

In this document, we define a new type of VDS, associated with the Confidential Consortium Framework (CCF) ledger. This VDS carries indexed transaction information in a binary Merkle Tree, where new transactions are appended to the right, so that the binary decomposition of the index of a transaction can be interpreted as the position in the tree if 0 represents the left branch and 1 the right branch.
Compared to {{-certificate-transparency-v2}}, the leaves of CCF trees carry additional internal information for the following purposes:

1. To bind the full details of the transaction executed, which is a super-set of what is exposed in the proof and captures internal information details useful for detailed system audit, but not for application purposes.

> 1. To verify that elements are only written by the Trusted Execution Environment, which addresses the persistence of committed transactions that happen between new signatures of the Merkle Tree root.

Reviewer 2: ```suggestion
2. To verify that elements are only written by the Trusted Execution Environment, which addresses the persistence of committed transactions that happen between new signatures of the Merkle Tree root.
```

informative:
  I-D.steele-cose-merkle-tree-proofs: COMTRE

  CCF:
    title: "Confidential Consortium Framework"
    target: "https://github.com/microsoft/ccf"

--- abstract

This document defines a new verifiable data structure type for COSE Signed Merkle Tree Proofs specifically designed for transaction ledgers produced by Trusted Execution Environments (TEEs), such as the Confidential Consortium Framework ({{CCF}}) to provide stronger tamper-evidence guarantees.

--- middle

# Introduction

> The Concise Encoding of Signed Merkle Tree Proofs (CoMeTre) {{-COMTRE}} defines a common framework for defining different types of proofs, such as proof of inclusion, about verifiable data structures (VDS). For instance, inclusion proofs guarantee to a verifier that a given serializable element is recorded at a given state of the VDS, while consistency proofs are used to establish that an inclusion proof is still consistent with the new state of the VDS at a later time.

Reviewer 3: Maybe:
```suggestion
The COSE Receipts document {{-COMTRE}} defines a common framework for defining different types of proofs, such as proof of inclusion, about verifiable data structures (VDS). For instance, inclusion proofs guarantee to a verifier that a given serializable element is recorded at a given state of the VDS, while consistency proofs are used to establish that an inclusion proof is still consistent with the new state of the VDS at a later time.
```
Also, shouldn't I-D.steele-cose-merkle-tree-proofs (`{{-COMETRE}}`)be a normative reference?

Reviewer 3: Err... I-D.**IETF**-cose-merkle-tree-proofs I mean, of course :sweat_smile:

Author: Will fix

--- abstract

This document defines a new verifiable data structure type for COSE Signed Merkle Tree Proofs specifically designed for transaction ledgers produced by Trusted Execution Environments (TEEs) to provide stronger tamper-evidence guarantees.

--- middle

# Introduction

> The Concise Encoding of Signed Merkle Tree Proofs (CoMeTre) {{-COMTRE}} defines a common framework for defining different types of proofs, such as proof of inclusion, about verifiable data structures (VDS). For instance, inclusion proofs guarantee to a verifier that a given serializable element is recorded at a given state of the VDS, while consistency proofs are used to establish that an inclusion proof is still consistent with the new state of the VDS at a later time.

Author: @OR13 I saw in #3 that you suggested adding references, which seems like an excellent idea. Do you have some pointers on the best way to add references in this kind of document? Just a link to the project's GitHub/paper?

Reviewer: Here is an example of how to add an arbitrary informative reference:

https://github.com/cose-wg/draft-ietf-cose-dilithium/pull/5/files#diff-2478aa1dfa824cbb11926553a0f4f8c9cd3e223126798f02e4827a8a56e809e9R62

- : denotes concatenation of lists
- D\[k1:k2\] = D'_(k2-k1) denotes the list \{d'\[0\] = d\[k1\], d'\[1\] = d\[k1+1\], ..., d'\[k2-k1-1\] = d\[k2-1\]\} of length (k2 - k1).

## Transaction Components

Each leaf in a CCF ledger carries the following components:

~~~
CCF-leaf = [

  ; a string of HASH_SIZE bytes
  internal-transaction-hash: bstr

  ; a string of at most 1024 bytes;
  internal-evidence: bstr

  ; a string of HASH_SIZE bytes
  data-hash: bstr
]
~~~

> The `internal-transaction-hash` and `internal-evidence` byte strings are internal to the CCF implementation. They can be safely ignored by receipt Verifiers, but they commit the TS to the whole tree contents and may be used for additional, CCF-specific auditing.

Reviewer 3: If internal-transaction-hash and internal-evidence are not defined here, but are specified in CCF, maybe a reference to the definition is appropriate here?

Author: I do not know to what extent it is ok to be specific, but I will add references.

# CCF Inclusion Proofs

CCF inclusion proofs consist of a list of digests tagged with a single left-or-right bit.

~~~

> CCF-inclusion-proof: [

Reviewer 3: labels/keys in arrays are intended to improve readability for humans and are not conveyed in the end. Was the intend maybe to use a map here, instead?

If so, is that also a possibility for other arrays that include elements with string label definitions (e.g., CCF-leaf that looks to me more like an annotation for humans that read the CDDL definition, but now I am not so sure)?

As all elements are mandatory, an array is an okay choice in any case. Just unsure about the intend of the labels.

Author: I think I prefer an array over a map here, repeating the keys seems quite wasteful considering this is not a self-describing document. I think I should have used comments here rather than labels, let me fix that.

Author: Sorry read that too quickly, a map does seem preferable at the top-level.

  leaf (label TBD_2): CCF-leaf
  path (label TBD_3): [+ ccf-proof-element]
]

ccf-proof-element = [

  ; position of the element
  left: bool

> 

Reviewer 3: ```suggestion
```
Why newlines? For readability? Maybe appending the comments behind the array members combines readability and no newlines in the array definition.

Author: I adopted the style I found in https://datatracker.ietf.org/doc/draft-ietf-cose-hash-envelope/, but I am super happy to line up on what's generally preferred.

Author: Reverted to something more compact.

- : denotes concatenation of lists
- D\[k1:k2\] = D'_(k2-k1) denotes the list \{d'\[0\] = d\[k1\], d'\[1\] = d\[k1+1\], ..., d'\[k2-k1-1\] = d\[k2-1\]\} of length (k2 - k1).

## Transaction Components

Each leaf in a CCF ledger carries the following components:

~~~
ccf-leaf = [
  internal-transaction-hash: bstr ; a string of HASH_SIZE bytes
  internal-evidence: bstr         ; a string of at most 1024 bytes

>   data-hash: bstr                 ; a string of HASH_SIZE bytes

Reviewer 2: Shall `data-hash` be explained below too?

Author: done

- : denotes concatenation of lists
- D\[k1:k2\] = D'_(k2-k1) denotes the list \{d'\[0\] = d\[k1\], d'\[1\] = d\[k1+1\], ..., d'\[k2-k1-1\] = d\[k2-1\]\} of length (k2 - k1).

## Transaction Components

Each leaf in a CCF ledger carries the following components:

~~~
ccf-leaf = [
  internal-transaction-hash: bstr ; a string of HASH_SIZE bytes

>   internal-evidence: bstr         ; a string of at most 1024 bytes

Author: ```suggestion
  internal-evidence: tstr         ; a string of at most 1024 bytes
```

  data-hash: bstr                 ; a string of HASH_SIZE bytes
]
~~~

The `internal-transaction-hash` and `internal-evidence` byte strings are internal to the CCF implementation. They can be safely ignored by receipt Verifiers, but they commit the TS to the whole tree contents and may be used for additional, CCF-specific auditing.

`internal-transaction-hash` is a hash over the complete entry in the {{CCF-Ledger-Format}}, and `internal-evidence` is a revealable {{CCF-Commit-Evidence}} value that allows early persistence of ledger entries before distributed consensus can be established.

`data-hash` summarises the subject of the proof: the data which is included in the ledger at this transaction.

# CCF Inclusion Proofs

CCF inclusion proofs consist of a list of digests tagged with a single left-or-right bit.

~~~

> ccf-inclusion-proof-map = {

Author: ```suggestion
ccf-inclusion-proof = bstr .cbor {
```

Closer to draft 5 of https://datatracker.ietf.org/doc/draft-ietf-cose-merkle-tree-proofs/

> ccf-inclusion-proof = bstr .cbor {

Author: To be validated and corrected as necessary

Author: ![image](https://github.com/user-attachments/assets/e6032ce8-fa08-401e-b1d1-feaf8e23d724)

The unprotected header for a CCF inclusion proof signature MUST include the following:

* `inclusion-proof: bstr .cbor ccf-inclusion-proof`. This contains the serialized CCF inclusion proof, as defined above.

The payload of the signature is the CCF ledger Merkle root digest, and MUST be detached in order to force verifiers to recompute the root from the inclusion proof in the unprotected header. This provides a safeguard against implementation errors that use the payload of the signature but do not recompute the root from the inclusion proof.

## Inclusion Proof Verification Algorithm

CCF uses the following algorithm to verify an inclusion receipt:

~~~
compute_root(proof):
  h := proof.leaf.internal-transaction-hash
       || HASH(proof.leaf.internal-evidence)
       || proof.leaf.data-hash

  for [left, hash] in proof:
      h := HASH(hash + h) if left
           HASH(h + hash) else
  return h

verify_inclusion_receipt(inclusion_receipt):
  let proof = inclusion_receipt.unprotected_headers[INCLUSION_PROOF_LABEL] or fail
  assert(inclusion_receipt.payload == nil)
  let payload = compute_root(proof)

  # Use the Merkle Root as the detached payload

>   return verif_cose(inclusion_receipt, payload)

Reviewer 2: ```suggestion
  return verify_cose(inclusion_receipt, payload)
```

* Value type: `bstr`
* Reference: This document

> * Label: TBD_3

Reviewer 3: Polishing the registration section is still a todo, but that does not have to be part of this PR.

Author: Sorry for asking a lot of dumb questions, but can you help me understand what's needed in that section, perhaps through a pointer to somewhere that does it right?

Author: To be done in separate PR.
